### ISSUE TITLE
Fardels grant #2 - roadmap to a sustainable dApp

### ISSUE THREAD

**Issue by darwinzer0** [SecretFoundation/Grants#21]
# Fardels grant 2 - roadmap to a sustainable dApp

The previous Fardels grant allowed us to successfully create a prototype version of the Fardels website that was launched and tested on holodeck-2. Since testing, we've added more functionality to the contract and are currently doing a second round of testing. Now that we are close to launching the mainnet version of the Fardels contract, we are asking for a new grant that will enable us to build this project into a sustainable startup. 

Our plan is to release the mainnet version of the contract late Aug/early Sept 2021. At that point the fardels website (UI) will be released in beta and development will continue on the website as well as other off-chain infrastucture for indexing. 

### Feature enchancements over the next 6 months (not exhaustive list). 

These are in rough order of priority, but many can be developed in parallel:

- Preparation for mainnet release of contract
    - Final testing of contract through remainder of August
    - Setting up web server infrastructure
- Full search functionality to discover fardels that match keywords / interests of the user. (See more below.)
- Building out application features for interacting with fardel contents within the app, especially encrypted files of different types that are stored on IPFS. To begin with we will focus on views for the content. Later on we can focus on building out capability for content creation directly in the website. File-based views include:
    - Rendering of images, 3D models
    - Playback of videos and audio
    - Readers for popular document types, including PDF, Word, etc.
- Integration of IPFS pinning services into file upload, starting with Pinata. IPFS settings can be stored as private user settings in the contract.
- Adding in option to create countable fardels (i.e. with maximum number of unpacks). The contract already supports this.
- Support for linking fardels and embedding in other websites and social media:
	- Shareable links to individual fardels and accounts.
	- Open source Javascript API for third-party applications to be able to upload fardels to the network, embed read-only queries, etc.
- Optimizations to improve UX, e.g., by caching contract queries in browser (can be enabled/disabled by user).
- Direct messaging with Altermail integration.
- UI controls to set the "approval required" flag in the fardel, which makes any unpack go into a "pending" status until it has been approved. This prevents a malicious node runner from unpacking fardels without paying for them (see https://forum.scrt.network/t/product-idea-secret-market/2961). By default this will not be set, because fardels are by definition low-value items and running a malicious node is expensive, so not really worth the trouble. It adds a delay on unpacking and an additional expense for the poster of the fardel, because each approval will cost gas fees. But the option is there for those who really want it--makes most sense in the context of countable fardels. Part of this will involve developing an open source bot that people can install that automatically queries for pending unpacks and approves them.
- Ability to browse transaction history (as seller or buyer), and download.
- Guided tutorials for new users and FAQ page.

### Indexing public data

The ability to search for fardels is critical--currently the interface only allows search by user handle. People can share their handle on other social media but clearly we need other ways for people to discover content. Design and implementation of an elasticsearch index for public data will be a high priority. Search functionality should match user profiles (handles, description, view settings, e.g. categories), fardel (public message, tags). In addition, faceted category-based search will be implemented.

### Design work and driving engagement

In parallel with feature development, we will fully upgrade the design of the Fardels website, including not only logo/branding but also streamlining the user interface to create an accessible and easy-to-use interface with a unique look and feel. The goal here is to create a website that will stand out and create a different experience from other social media, and one that visually evokes the metaphor of carrying a fardel to Secret Network.

Once the contract is on mainnet, we will reward early engagement by creating a set of secretNFTs to be given to users who first reach milestones on the social network, e.g.: first with (20, 100, etc.) followers, first with (20, 100, etc.) unpacks, etc. In terms of growing the user base, we will leverage the Secret Network community as best we can but also will be looking to engage with content creators, focusing in a few areas such as music, art, and design. We'd prefer to start at a relatively slow burn, developing as lean as possible while learning from user feedback about what features to prioritize. Our long term goal is that the impetus for people to use the site will be that it offers a decentralized mechanism for people to share and monetize their content, and the fact that it is built on Secret Network will just be an implementation detail. 

### Longer-term development goals

There are a few longer-term development goals that are probably out-of-scope for this funding timeline, but we might get to depending on how things develop. 

- Mobile version. This will likely not happen until there are good mobile wallet APIs for executing secret network contracts. A progressive web app might be the best approach, though our team also has extensive native app development experience so we will weigh all options.
- Content creation modules integrated into the app. Rather than uploading files or links to finished content, the app could support content creation directly within the app (e.g., substack-like blogging or video recording). However, we need to understand the use cases better, which will be clearer after getting user feedback.
- Currently files can only be stored on IPFS as single files. This means that a full download must happen before playback of video, for example. Future work will explore uploading multiple files to support features like streaming or the uploading of directories.
- Features that will require an update to the contract:
	- Allow for payment using SNIP-20 tokens. This will give two key advantages: 1) all transaction amounts will be hidden, 2) can allow for pricing a fardel using a fiat tethered coin (e.g. sUSDT or sUST after terra bridge). 
	- Developing mechanisms that remove the disincentivize to comment and rate fardels, by refunding gas fees. This might require holding some percentage of the transaction fee in the contract.

### Timeline and funding request

The timeline for this grant is envisioned to last 6 months, with 3 payment installments 1) early Sept (after mainnet release), 2) ~Nov 1, 2021 (mid-term report), and 3) ~Feb 1, 2022 (final report). The funding we are asking for is:

(10,000 SCRT / installment) x 3 = 30,000 SCRT total. 

Payment of the grant will be divided 84% to US-based Fardels, LLC (8400 SCRT / installment) and 16% (1600 SCRT / installment) to Tim Denee (design lead).

The payment to Fardels, LLC will help to cover all dev personnel working on the project as well as some modest operating expenditures, which consist primarily of cloud computing to host the Fardels web server (plus load balancing, etc.) and a server for hosting the elasticsearch index. 

Once the mainnet contract is live we anticipate that user feedback might shift some priorities, and we want the ability to flexibly respond to those needs as they arise. We will provide status updates to demonstrate progress on the project for each installment (in addition, to more frequent, smaller updates on discord and twitter, etc). 

Team leads: 
- Ben Adams (project and programming lead)
- Tim Denee (design lead) https://www.timdenee.com

**Comment by darwinzer0** [comment]
Most of the listed feature enhancements have already been implemented, so closing this funding request.
